File: app/src/main/java/com/jeketos/associatewith/drawer/DrawerActivity.java
package com.jeketos.associatewith.drawer;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.DisplayMetrics;
import android.widget.ImageView;
import com.jeketos.associatewith.Point;
import com.jeketos.associatewith.R;
import com.jeketos.associatewith.di.Injector;
import com.jeketos.associatewith.guesser.chat.ChatAdapter;
import com.jeketos.associatewith.guesser.chat.IChatItem;
import com.jeketos.associatewith.listener.TouchListener;
import com.jeketos.associatewith.util.DialogUtils;

import butterknife.BindView;
import butterknife.ButterKnife;

public class DrawerActivity extends AppCompatActivity implements DrawerMVP.DrawerView {

//    private static final String TAG = "DrawerActivity";
    private Canvas canvas;
    private Paint paint;
    @BindView(R.id.image_view)
    ImageView imageView;
    DrawerMVP.DrawerPresenter presenter;
    @BindView(R.id.recycler_view)
    RecyclerView chatRecyclerView;

    private TouchListener.MoveListener moveListener = new TouchListener.MoveListener() {
        @Override
        public void actionDown(Point point) {
            drawPoint(point);
        }

        @Override
        public void actionMove(float previousX, float previousY, Point point) {
            drawLine(previousX, previousY, point);
        }

        @Override
        public void actionUp(float previousX, float previousY, Point point) {
            drawLine(previousX, previousY, point);
        }
    };
    private TouchListener onTouchListener;
    private ChatAdapter chatAdapter;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_drawer);
        ButterKnife.bind(this);
        onTouchListener = new TouchListener(moveListener);
        presenter = Injector.INSTANCE.provideDrawPresenter(this);

    }

    @Override
     public void init() {
        DisplayMetrics displayMetrics = getResources().getDisplayMetrics();
        Bitmap bitmap = Bitmap.createBitmap(displayMetrics.widthPixels,displayMetrics.heightPixels,
                Bitmap.Config.ARGB_8888);
        canvas = new Canvas(bitmap);
        paint = new Paint();
        paint.setColor(Color.BLACK);
        paint.setAntiAlias(true);
        paint.setDither(true);
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeJoin(Paint.Join.ROUND);
        paint.setStrokeCap(Paint.Cap.ROUND);
        paint.setStrokeWidth(12);
        imageView.setImageBitmap(bitmap);
        imageView.setOnTouchListener(onTouchListener);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this);
        chatRecyclerView.setLayoutManager(layoutManager);
        chatAdapter = new ChatAdapter();
        chatRecyclerView.setAdapter(chatAdapter);
    }

    @Override
    public void addChatItem(IChatItem chatItem) {
        chatAdapter.updateItems(chatItem);
    }

    @Override
    public void clearChat() {
        chatAdapter.updateItems(null);
    }

    @Override
    public void showChooseWordDialog(CharSequence[] words) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(R.string.choose_word);
        builder.setCancelable(false);
        builder.setItems(words, (dialogInterface, i) -> {
            presenter.saveWord(words[i]);
            dialogInterface.dismiss();
        });
        builder.create().show();
    }

    @Override
    public void showWinnerDialog(String name) {
        AlertDialog.Builder builder = DialogUtils.createWinnerDialog(this,name,null);
        builder.create();
        builder.show();
    }


    public void drawPoint(Point point) {
        canvas.drawPoint(point.getX(), point.getY(), paint);
        imageView.invalidate();
        presenter.sendPoint(point);
    }


    public void drawLine(float previousX, float previousY, Point point) {
        canvas.drawLine(previousX, previousY, point.getX(), point.getY(), paint);
        imageView.invalidate();
        presenter.sendPoint(point);
    }
}

File: app/src/main/java/com/jeketos/associatewith/drawer/DrawerActivity.kt
package com.jeketos.associatewith.drawer

import android.content.Intent
import android.os.Bundle
import android.support.design.widget.BottomSheetBehavior
import android.support.v4.view.GravityCompat
import android.support.v7.app.AlertDialog
import android.support.v7.widget.LinearLayoutManager
import com.afollestad.materialdialogs.color.ColorChooserDialog
import com.jeketos.associatewith.Point
import com.jeketos.associatewith.R
import com.jeketos.associatewith.base.BaseActivity
import com.jeketos.associatewith.base.BasePresenter
import com.jeketos.associatewith.chat.ExtChatAdapter
import com.jeketos.associatewith.chat.IChatItem
import com.jeketos.associatewith.guesser.GuesserActivity
import com.jeketos.associatewith.listener.TouchWatcher
import com.jeketos.associatewith.util.DialogUtils
import kotlinx.android.synthetic.main.activity_drawer.*
import kotlinx.android.synthetic.main.drawer_layout.*
import kotlinx.android.synthetic.main.drawer_picker.*
import javax.inject.Inject

class DrawerActivity() : BaseActivity<DrawerMVP.DrawerView>(), DrawerMVP.DrawerView, ColorChooserDialog.ColorCallback {


    @Inject lateinit var presenter : DrawerMVP.DrawerPresenter
    lateinit var extChatAdapter: ExtChatAdapter

    val touchWatcher = object : TouchWatcher {

        override fun actionTouch(point: Point) {
            presenter.sendPoint(point)
        }
    }

    override fun getPresenter(): BasePresenter<DrawerMVP.DrawerView> {
        return presenter
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        component.inject(this)
        setContentView(R.layout.drawer_layout)
        showProgressDialog()
        init()
    }

    override fun onStart() {
        super.onStart()
        presenter.init()
    }

    override fun onStop() {
        super.onStop()
    }

    override fun init() {
        imageView.setMoveWatcher(touchWatcher)
        val layoutManager = LinearLayoutManager(this)
        chatRecyclerView.layoutManager = layoutManager
        layoutManager.reverseLayout = true
        layoutManager.stackFromEnd = true
        extChatAdapter = ExtChatAdapter(this)
        chatRecyclerView.adapter = extChatAdapter
        drawerImage.setOnClickListener {
            drawerLayout.openDrawer(GravityCompat.END)
        }
        clearButton.setOnClickListener {
            imageView.clear()
            presenter.clearDraw()
            drawerLayout.closeDrawer(GravityCompat.END)
        }
        chooseColorButton.setOnClickListener {
            ColorChooserDialog.Builder(this, R.string.change_color)
                    .titleSub(R.string.color_shades)  // title of dialog when viewing shades of a color
                    .accentMode(false)  // when true, will display accent palette instead of primary palette
                    .doneButton(R.string.md_done_label)  // changes label of the done button
                    .cancelButton(R.string.md_cancel_label)  // changes label of the cancel button
                    .backButton(R.string.md_back_label)  // changes label of the back button
                    .dynamicButtonColor(true)  // defaults to true, false will disable changing action buttons' color to currently selected color
                .show()
            drawerLayout.closeDrawer(GravityCompat.END)
        }
        strokeRecyclerView.layoutManager = LinearLayoutManager(this)
        strokeRecyclerView. adapter = StrokeWidthAdapter(this)
        val bottomSheetBehavior = BottomSheetBehavior.from(bottomSheet)
        bottomSheetBehavior.peekHeight = resources.getDimension(R.dimen.bottom_sheet_peek).toInt()
        bottomSheetBehavior.state = BottomSheetBehavior.STATE_COLLAPSED
    }

    override fun onColorSelection(dialog: ColorChooserDialog, selectedColor: Int) {
        imageView.setDrawColor(selectedColor)
    }

    override fun addChatItem(chatItem: IChatItem?) {
        extChatAdapter.updateItems(chatItem)
    }

    override fun clearChat() {
        extChatAdapter.updateItems(null)
    }

    override fun setStrokeWidth(strokeWidth: Float) {
        imageView.setStrokeWidth(strokeWidth)
        drawerLayout.closeDrawer(GravityCompat.END)
    }

    override fun updateChatItemColor(chatItem: IChatItem, position: Int) {
        presenter.updateChatItemColor(chatItem, position)
    }

    override fun showChooseWordDialog(words: Array<CharSequence>) {
        hideProgressDialog()
        val builder = AlertDialog.Builder(this)
        builder.setTitle(R.string.choose_word)
        builder.setCancelable(false)
        builder.setItems(words, { dialogInterface, i ->
            presenter.saveWord(words[i])
            dialogInterface.dismiss()
        })
        builder.create().show()
    }

    override fun showWinnerDialog(name: String) {
        val builder = DialogUtils.createWinnerDialog(this,name,null)
        builder.setPositiveButton(android.R.string.ok, {d,i ->
            finish()
            startActivity(Intent(this, GuesserActivity::class.java))
        })
        builder.create()
        if(!this.isFinishing)
            builder.show()
    }
}
